### PR TITLE
Type updates from @mr-c - part 2

### PR DESCRIPTION
**Pull request by mr-c**
Why: common-workflow-lab#137
What: Improved type hints

## How to test the changes?
(Select all options that apply)
- [x] This is a refactoring of components with existing test coverage.

## License
- [x] I agree to license these contributions under [Galaxy's current license](https://github.com/galaxyproject/galaxy/blob/dev/LICENSE.txt).
- [x] I agree to allow the Galaxy committers to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT). If this condition is an issue, uncheck and just let us know why with an e-mail to galaxy-committers@lists.galaxyproject.org.

### CODE REVIEW

File: lib/galaxy/datatypes/registry.py [hunk 1]
```diff
@@ -6,11 +6,12 @@
 import logging
 import os
 from string import Template
-from typing import Dict
+from typing import Dict, List, Optional, Tuple
 
 import yaml
 
 import galaxy.util
+from galaxy.model import DatasetInstance
```
Comment (mr-c) [inline lib/galaxy/datatypes/registry.py]: TODO: move this under `TYPE_CHECKING`

**Comment by mr-c** [comment]
To wait until after #12868

File: lib/galaxy/tools/actions/__init__.py [hunk 2]
```diff
@@ -51,12 +51,12 @@ class ToolAction:
     been converted and validated).
     """
 
-    def execute(self, tool, trans, incoming=None, set_output_hid=True):
+    def execute(self, tool, trans, incoming=None, set_output_hid=True, **kwargs):
         incoming = incoming or {}
```
Comment (nsoranzo) [inline lib/galaxy/tools/actions/__init__.py]:
```suggestion
```

Also, should this be decorated `@abstractmethod` instead of raising `TypeError` ?

Comment (mr-c) [reply lib/galaxy/tools/actions/__init__.py]: If I add `@abstractmethod` and inherit from `ABCMeta` then:
```
lib/galaxy/tools/__init__.py:993: error: All overload variants require at least one argument  [call-overload]
lib/galaxy/tools/__init__.py:993: note: Possible overload variants:
lib/galaxy/tools/__init__.py:993: note:     def (o: object) -> DefaultToolAction
lib/galaxy/tools/__init__.py:993: note:     def (name: str, bases: Tuple[type, ...], dict: Dict[str, Any], **kwds: Any) -> DefaultToolAction
test/unit/app/tools/test_actions.py:72: error: All overload variants of "DefaultToolAction" require at least one argument  [call-overload]
test/unit/app/tools/test_actions.py:72: note: Possible overload variants:
test/unit/app/tools/test_actions.py:72: note:     def DefaultToolAction(self, o: object) -> DefaultToolAction
test/unit/app/tools/test_actions.py:72: note:     def DefaultToolAction(self, name: str, bases: Tuple[type, ...], dict: Dict[str, Any], **kwds: Any) -> DefaultToolAction

```

File: lib/galaxy/managers/collections_util.py [hunk 3]
```diff
@@ -111,7 +111,7 @@ def dictify_dataset_collection_instance(dataset_collection_instance, parent, sec
         encoded_history_id = security.encode_id(parent.id)
         dict_value['url'] = web.url_for('history_content_typed', history_id=encoded_history_id, id=encoded_id, type="dataset_collection")
     elif isinstance(parent, model.LibraryFolder):
-        encoded_library_id = security.encode_id(parent.library.id)
+        encoded_library_id = security.encode_id(parent.library.id)  # type: ignore
```
Comment (mr-c) [inline lib/galaxy/managers/collections_util.py]: I have no idea how this is supposed to work

File: lib/galaxy/managers/collections.py [hunk 4]
```diff
@@ -592,18 +599,23 @@ def _build_elements_from_rule_data(self, collection_type_description, rule_set,
...
                         sub_collection["collection_type"] = collection_type_at_depth.collection_type
                         sub_collection["elements"] = {}
-                        elements_at_depth[identifier] = sub_collection
-                        elements_at_depth = sub_collection["elements"]
+                        elements_at_depth[
+                            identifier
+                        ] = sub_collection  # ??? but this is overridenn on the next line??
+                        elements_at_depth = sub_collection[
+                            "elements"
+                        ]  # ??? sub_collection is just lost?
```
Comment (mr-c) [inline lib/galaxy/managers/collections.py]: Yes, this is a cry for help

File: lib/galaxy/tools/__init__.py [hunk 5]
```diff
@@ -664,6 +664,8 @@ def _view(self):
 
     @property
     def version_object(self):
+        if not self.version:
+            raise Exception("Must call parse() first")
```
Comment (nsoranzo) [inline lib/galaxy/tools/__init__.py]: `parse()` is called by `__init__()` ?

Comment (mr-c) [reply lib/galaxy/tools/__init__.py]: Agreed, but without this we get https://app.circleci.com/pipelines/github/galaxyproject/galaxy/19099/workflows/84a60b86-6e6b-46f3-909b-62844104652b/jobs/156302?invite=true#step-104-2387

> `galaxy/tools/__init__.py:667: error: Argument 1 to "parse" has incompatible type "Optional[Any]"; expected "str"  [arg-type]`

File: lib/galaxy/managers/collections.py (hunk 4 above)
Comment (mvdbeek) [reply lib/galaxy/managers/collections.py]: Yeah, I think all of this (L602-L611) could be replaced with `elements_at_depth = {}`, and we'll see if any tests break.

**Comment by jmchilton** [comment]
Amazing - thanks so much!

**Pull request merged**